File: Recursion/Leetcode/Medium/GenerateParenthesis/GenerateParenthesis.cpp
// https://leetcode.com/problems/generate-parentheses/
#include<bits/stdc++.h>
using namespace std;
void generateP(string s, int n, vector<string> &v, int left, int right)
{
    if (s.length() == n * 2)
    {
        v.push_back(s);
        return;
    }
    if (left < n)
    {
        generateP(s + "(", n, v, left + 1, right);
    }
    if (right < left)
    {
        generateP(s + ")", n, v, left, right + 1);
    }
}
vector<string> generateParenthesis(int n)
{
    vector<string> s;
    generateP("", n, s, 0, 0);
    return s;
}
int main(void) {
    vector<string> str = generateParenthesis(3);
    for(string ele: str) {
        cout << ele << " ";
    }
    cout << endl;
    return 0;
}

File: Recursion/Leetcode/Medium/GenerateParenthesis/GenerateParenthesis.java
// https://leetcode.com/problems/generate-parentheses/
import java.util.ArrayList;
import java.util.List;

public class GenerateParenthesis {
    static void generate(int n, String str, int left, int right, List<String> combo) {
        if (str.length() == 2 * n) {
            combo.add(str);
            return;
        }
        if (left < n) {
            generate(n, str + "(", left + 1, right, combo);
        }
        if (right < left) {
            generate(n, str + ")", left, right + 1, combo);
        }
    }

    static List<String> generateParenthesis(int n) {
        List<String> v = new ArrayList<>();
        generate(n, "", 0, 0, v);
        return v;
    }
    public static void main(String[] args) {
        List<String> v = generateParenthesis(3);
        System.out.println(v);
    }
}
